lib/shelley/src/Cardano/Wallet/Shelley/Pools.hs

```diff
@@ -517,31 +525,52 @@ monitorStakePools tr gp nl db@DBLayer{..} = do
                             liftIO $ traceWith tr $ MsgErrProduction e
                         Right () ->
                             pure ()
-
-                -- A single block can contain multiple certificates relating to the
-                -- same pool.
-                --
-                -- The /order/ in which certificates appear is /significant/:
-                -- certificates that appear later in a block /generally/ take
-                -- precedence over certificates that appear earlier on.
-                --
-                -- We record /all/ certificates within the database, together with
-                -- the order in which they appeared.
-                --
-                -- Precedence is determined by the 'readPoolLifeCycleStatus'
-                -- function.
-                --
-                let publicationTimes =
-                        CertificatePublicationTime slot <$> [minBound ..]
-                forM_ (publicationTimes `zip` certificates) $ \case
-                    (publicationTime, Registration cert) -> do
-                        liftIO $ traceWith tr $ MsgStakePoolRegistration cert
-                        putPoolRegistration publicationTime cert
-                    (publicationTime, Retirement cert) -> do
-                        liftIO $ traceWith tr $ MsgStakePoolRetirement cert
-                        putPoolRetirement publicationTime cert
+                unless (null certificates) $ do
+                    -- Before adding new pool certificates to the database, we
+                    -- first attempt to garbage collect pools that have already
+                    -- retired. By only performing garbage collection when we
+                    -- actually have new certificates to add, we avoid paying
+                    -- the cost of garbage collection in every single slot.
+                    garbageCollectPools slot
+                    putPoolCertificates slot certificates
         pure Continue
 
+    -- Perform garbage collection for pools that have retired.
+    --
+    -- To avoid any issues with rollback, we err on the side of caution and
+    -- only garbage collect pools that retired /at least/ two epochs before
+    -- the current epoch.
+    --
+    garbageCollectPools currentSlot = liftIO $ do
+        currentEpoch <- timeInterpreter nl (epochOf currentSlot)
+        let subtractTwoEpochs = epochPred <=< epochPred
```

Comment:
What if the current epoch is 0? (Perhaps not technically possible this is called then, but still…)

Reply:
> What if the current epoch is 0? (Perhaps not technically possible this is called then, but still…)

I'm glad you asked :D

In that case, we will not perform any garbage collection, as `subtractTwoEpochs currentEpoch` will evaluate to `Nothing`. (Since [`epochPred`](https://github.com/input-output-hk/cardano-wallet/blob/fece720cb18d9676c4f1f6b0d3a69a37fab93f1c/lib/core/src/Cardano/Wallet/Primitive/Slotting.hs#L364) has type `EpochNo -> Maybe EpochNo`)